CMake: Use armclang for ASM files

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

The ARM Assembler `armasm` has a bug processing macro with array values. It omits the `,` delimiter which results in syntax errors.
As it is being deprecated, Arm Compiler `armclang` is the recommended tool to compile ARM Assembly Language.
However, given that the various Mbed ARM TOOLCHAIN assembly code are written using the legacy `armasm` syntax, the command line option `-masm` needs to be passed to `armclang` so it can select the correct assembler for the input assembly source files. This has been successfully tested with `mbed-os-example-blinky`.
The option was set to `auto` as there are circumstances where the auto-detection might select the wrong assembler for the input file. One of thos circumstances is if the input file is a .S file that requires preprocessing, this is the case for a number of Mbed Assembly source file.
Note: The current release of CMake (3.18.2) has a bug for `armclang` where it does not append target definitions to the ASM compiler command.
This [was reported](https://gitlab.kitware.com/cmake/cmake/-/issues/21148), subsequentenly [fixed](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5175) and scheduled to be part of the next CMake release.
Thanks @hugueskamba for fixing it.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
CMake ARMClang6: use armclang for asm

Fixes the issue of armasm, passing -DMACRO={0,0} - fails to parse.

We use armclang frontend but passing -masm=armasm.

0xc0170 committed:
CMake cores: fix flag  mpcu for asm

0xc0170 commented on cmake/toolchains/ARM.cmake:
```diff
@@ -42,14 +42,10 @@ function(mbed_set_toolchain_options target)
-    )
     target_compile_options(${target}
         PUBLIC
+            $<$<COMPILE_LANGUAGE:ASM>:--target=arm-arm-none-eabi -masm=armasm>
```
`-masm=auto` works for me. If confirmed, I'll change this

hugueskamba replied:
`-masm=auto` also worked on my Mac OS system with `cmake version 3.17.3`.
However, would it not be prudent to explicitly set it?
```
In rare circumstances, the auto-detection might select the wrong assembler for the input file.
For example, if the input file is a .S file that requires preprocessing, auto-detection might select
the wrong assembler. For the files where auto-detection selects the wrong assembler, you must
select -masm=gnu or -masm=armasm explicitly.
```
Source: https://www.keil.com/support/man/docs/armclang_ref/armclang_ref_yml1538053827173.htm

0xc0170 replied:
For now, armasm would be better but what about the future? New MCU startup files will soon have GNU syntax for ARMClang .S files, therefore this armasm won't work. Auto mode is neat and exactly what we need. if we find bugs, we shall report it to the compiler (it is state as "rare").

I would rather take  care of these rare cases (overwrite armasm to use gnu manually - the component would request to be compiled with GNU syntax for ARMClang6, for instance target component for specific board). Is this feasible ?

LDong-Arm commented on cmake/cores/Cortex-M0+.cmake:
```diff
@@ -27,7 +27,7 @@ function(mbed_set_cpu_core_options target mbed_toolchain)
                 $<$<COMPILE_LANGUAGE:C>:${compile_options}>
                 $<$<COMPILE_LANGUAGE:CXX>:${compile_options}>
-                $<$<COMPILE_LANGUAGE:ASM>:--cpu=Cortex-M0plus>
+                $<$<$<$<COMPILE_LANGUAGE:ASM>:-mcpu=Cortex-M0plus>
```
I'm not a CMake expert but the extra `$<$<` looks imbalanced?

0xc0170 replied:
Good catch, fixing now

0xc0170 committed:
CMake ARMClang: use -masm=auto

Auto mode for ASM files. There might be rare cases where this fails, we will fix
them by requesting the component to use gnu syntax for assembly files.
